text/0456-elasticache-l2.md

```diff
@@ -0,0 +1,375 @@
+session stores, gaming leaderboards, streaming, and analytics. ElastiCache is
+compatible with Redis and Memcached.
+
+This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aws-cdk)
+project. It allows you to define ElastiCache Redis replication groups.
+
+## Constructs
+
+### Defining a RedisReplicationGroup
+
+In order to define a `RedisReplicationGroup` (Redis (cluster mode disabled)),
+you must specify the vpc to run the instance in.
+
+```ts
+import * as elasticache from '@aws-cdk/aws-elasticache';
+import * as ec2 from '@aws-cdk/aws-ec2';
+
+const vpc = new ec2.Vpc(this, 'Vpc');
+new elasticache.RedisReplicationGroup(this, "RedisReplicationGroup", {
+  engineVersion: elasticache.RedisEngineVersion.VER_6_2,
+  vpc,
+});
+```
+
+The above example implicitly defines the following resources:
+
+- A `VPC` with private subnets.
+- A `SubnetGroup` for all private subnets of the Vpc.
+- A `SecurityGroup` in the Vpc.
+- A `ReplicationGroup` with the security group attached, one shard with a Redis
+  engine version 6.2 of size `cache.t2.micro`.
+
+### Defining a RedisClusterReplicationGroup
+
+A `RedisClusterReplicationGroup` is similar to a `RedisReplicationGroup`, but
+with multiple shards. The documentation calls them "Redis (cluster mode enabled)".
+The main difference is that a `RedisClusterReplicationGroup` requires a
+`numNodeGroups` to be set to a value of 2 or higher. Only `RedisClusterReplicationGroup`s
+must have `automaticFailoverEnabled` enabled and exposes a `configurationEndpoint`,
+but cannot have a `snapshottingClusterId` set.
+
+### Other constructs
+
+The following resources are used as L1 constructs in the L2 constructs proposed here,
+but will not be exposed themselves:
+
+- SubnetGroup: Created based on the given vpcs subnets
+- ParameterGroup: Created based on the engineVersion and whether cluster-mode is enabled
+
+This RFC does not yet cover:
+
+- CacheCluster
+- GlobalReplicationGroup
+- SecurityGroup
+- User
+- UserGroup
+
+## Metrics
+
+ElastiCache provides metrics that enable you to monitor your clusters.
+You can access these metrics through CloudWatch. ElastiCache provides both
+host-level metrics (for example, CPU usage) and metrics that are specific to
+the cache engine software (for example, cache gets and cache misses).
+These metrics are measured and published for each Cache node in 60-second
+intervals.
+
+CDK provides methods for accessing metrics with default configuration,
+such as `metricCPUUtilization`, and `metricMemory`. CDK also provides a generic
+`metric` method that can be used to produce metric configurations for any [metric
+provided by Amazon ElastiCache for Redis](https://docs.aws.amazon.com/AmazonElastiCache/latest/red-ug/CacheMetrics.Redis.html).
+
+## Granting application access to a RedisReplicationGroup
+
+While ElastiCache manages rights to change the ElastiCache infrastructure via
+IAM, it does not implement read and write permission management the same way.
```

Comment:
Just wanna note that [ElastiCache supports IAM authentication](https://aws.amazon.com/about-aws/whats-new/2022/11/amazon-elasticache-supports-iam-authentication-redis-clusters/) since last Nov.

Reply:
Is this the recommended way to handle this now? If so, we should bake it into the L2 construct and, instead of managing auth via a Secret, add a `grantRead` method.

Reply:
I remember now - this is limited to Redis 7+. I would then not add this to the current construct, but potentially add a separate construct that only accepts engine Redis 7+ or higher, but offers a `grantRead` method. Otherwise, it would be confusing why `grantRead` sometimes works and sometimes doesn't, and might only error on synth time, instead of erroring already on compile time.

Reply:
I think I would be fine with throwing an error at synth time if the version is not supported. I'm not sure about managing different versions of the constructs for different versions of redis. It would make it hard to upgrade for one.

Reply:
Upgrading then should actually _just work_ by changing the version and the name of the used construct. I haven't seen that pattern in CDK yet, though, so if it doesn't exist yet, this would not be the best place to introduce it.

Reply:
Ok, then I would leave the RFC as is and leave the potential to add `grantRead` in the future?